packages/foundry/contracts/prediction-market/PredictionMarketHook.sol

```diff
@@ -0,0 +1,488 @@
+     * @param marketId Identifier of the market to swap
+     * @param from Side to swap from
+     * @param amountIn Amount to swap
+     * @return position Updated user position after the swap
+     */
+    function swap(
+        bytes32 marketId,
+        Side from,
+        uint256 amountIn
+    ) public returns (Position memory position) {
+        // get output amount from the market and prepare amount deltas
+        uint256 amountOut = markets.swap(marketId, from, amountIn, FEE);
+
+        int256 amountInDelta = amountIn.toInt256()*-1;
+        int256 amountOutdelta = amountOut.toInt256();
+
+        (int256 bullAmountDelta, int256 bearAmountDelta) = from == Side.Bull ? 
+            (amountInDelta, amountOutdelta) :
+            (amountOutdelta, amountInDelta);
+
+        // update the position with the amount deltas
+        Position memory updatedPosition = positions.applyPositionDelta(marketId, msg.sender, bullAmountDelta, bearAmountDelta);
+
+        return updatedPosition;
+    }
+
+    /**
+     * @notice Determines whether a given pool and token combination is valid
+     * @dev pool must contain both tokens in the pair to be valid. Otherwise we will not be able to price markets and 
+     * determine outcomes
+     * @param pool Pool hosting the prediction market
+     * @param tokenA First token in the pair
+     * @param tokenB Second token in the pair
+     * @return canCreateMarket True if the market can be created from the given pool and token
+     */
+    function _canCreateMarketFromPoolAndTokens(
+        address pool,
+        address tokenA,
+        address tokenB
+    ) private view returns (bool canCreateMarket) {
+        // cannot create markets with the same token
+        if(tokenA == tokenB) {
+            return false;
+        }
+
+        // get the pool tokens from the vault and return true if both tokens in the pair are contained in the pool
+        IERC20[] memory tokens = _vault.getPoolTokens(pool);
+        bool hasTokenA;
+        bool hasTokenB;
+
+        for(uint256 i=0; i<tokens.length;++i){
+            address token = address(tokens[i]);
+
+            if(token == tokenA){
+                hasTokenA = true;
+            }
+
+            if(token == tokenB) {
+                hasTokenB = true;
+            }
+        }
+
+        return hasTokenA && hasTokenB;
+    }
+
+    /**
+     * @notice Dontate fees back to pool LPs
+     * @param pool Pool to donate fees to
+     * @param token Fee denomination token
+     * @param amount Amount to donate
+     */
+    function _donate(
+        address pool,
+        address token,
+        uint256 amount
+    ) private {
+        IERC20[] memory tokens = _vault.getPoolTokens(pool);
+        uint256[] memory feeAmounts = new uint256[](tokens.length);
+
+        for(uint256 i=0; i<tokens.length;i++){
+            feeAmounts[i] = address(tokens[i]) == token ? amount : 0;
```

Comment:
optimization nit: `feeAmounts` is already initialized to all zeros; could just check if tokens[i] == token, set feeAmounts[i] and break.